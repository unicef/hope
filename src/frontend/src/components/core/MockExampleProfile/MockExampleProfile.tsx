import React from 'react';
import { useQuery } from '@tanstack/react-query';
import { RestService } from '@restgenerated/services/RestService';
import { useBaseUrl } from '@hooks/useBaseUrl';
import { Profile } from '@restgenerated/models/Profile';

const MockExampleProfile = () => {
  const { businessArea, programId } = useBaseUrl();

  const { data: meData, isLoading: meLoading } = useQuery<Profile>({
    queryKey: ['profile', businessArea, programId],
    queryFn: () => {
      return RestService.restBusinessAreasUsersProfileRetrieve({
        businessAreaSlug: businessArea,
        program: programId,
      });
    },
  });

  return (
    <div style={{ border: '3px solid black', padding: '10px', width: '200px' }}>
      {meLoading ? (
        'Loading...'
      ) : (
        <div>
          <h4>Profile Details</h4>
          {meData?.firstName} {meData?.lastName}
        </div>
      )}
    </div>
  );
};

export default MockExampleProfile;
