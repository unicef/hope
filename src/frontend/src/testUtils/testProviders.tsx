import { ThemeProvider } from '@mui/material';
import CssBaseline from '@mui/material/CssBaseline';
import * as React from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { theme } from '../theme';
import { AdapterMoment } from '@mui/x-date-pickers/AdapterMoment';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';

interface TestProvidersProps {
  children: React.ReactNode;
}

export const TestProviders: React.FC<TestProvidersProps> = ({ children }) => (
  <ThemeProvider theme={theme}>
    <StyledThemeProvider theme={theme}>
      <LocalizationProvider dateAdapter={AdapterMoment}>
        <CssBaseline />
        {children}
      </LocalizationProvider>
    </StyledThemeProvider>
  </ThemeProvider>
);
