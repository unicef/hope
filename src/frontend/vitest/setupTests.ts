import { random } from 'lodash';
import React from 'react';
import { setupServer } from 'msw/node';
import setupInternalization from '../src/i18n';
import * as useBusinessAreaModule from '../src/hooks/useBusinessArea';
import * as useGlobalProgramModule from '../src/hooks/useGlobalProgram';
import { handlers } from '../src/mocks/handlers';
import { beforeAll, afterAll, afterEach, beforeEach, vi } from 'vitest';

global.React = React;
// Set up global mocks and configurations
global.Date.now = () => new Date('1970-01-01T00:00:00.000Z').getTime();
process.env.TZ = 'UTC';
setupInternalization();

beforeEach(() => {
  vi.spyOn(global.Math, 'random').mockImplementation(random);
  vi.spyOn(useBusinessAreaModule, 'useBusinessArea').mockReturnValue(
    'afghanistan',
  );
  vi.spyOn(useGlobalProgramModule, 'useGlobalProgram').mockReturnValue(
    'UHJvZ3JhbU5vZGU6ZTRmOGMwNjctNjcwOC00NjZmLWFjYmMtZGE2OTkxZjE0MjY2',
  );
});

// Mock ResizeObserver for tests that rely on it
globalThis.ResizeObserver = class {
  observe() {}

  unobserve() {}

  disconnect() {}
};

// Mock crypto.randomUUID if not available in Vitest environment
globalThis.crypto.randomUUID = () => 'd7a794d1-0ead-4424-9ff2-38d14db32b99';

// Mock Service Worker registration
const server = setupServer(...handlers);

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that are declared as a part of our tests
// (i.e. for testing one-time error scenarios).
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());
