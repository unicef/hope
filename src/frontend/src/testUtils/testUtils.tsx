import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, RenderOptions } from '@testing-library/react';
import { Formik } from 'formik';
import noop from 'lodash/noop';
import { TestProviders } from './testProviders';
import { ProgramStatus } from '@generated/graphql';
import { MockLink } from '@apollo/client/testing';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) =>
  render(
    <>
      <BrowserRouter>
        <Formik initialValues={{}} onSubmit={noop}>
          {ui}
        </Formik>
      </BrowserRouter>
    </>,
    {
      wrapper: TestProviders,
      ...options,
    },
  );

export * from '@testing-library/react';
export { customRender as render };

let m_w = 123456789;
let m_z = 987654321;
const mask = 0xffffffff;

// Takes any integer
export const seed = (i) => {
  m_w = (123456789 + i) & mask;
  m_z = (987654321 - i) & mask;
};

// Returns number between 0 (inclusive) and 1.0 (exclusive),
// just like Math.random().
export const random = () => {
  m_z = (36969 * (m_z & 65535) + (m_z >> 16)) & mask;
  m_w = (18000 * (m_w & 65535) + (m_w >> 16)) & mask;
  let result = ((m_z << 16) + (m_w & 65535)) >>> 0;
  result /= 4294967296;
  return result;
};

export class ApolloLoadingLink extends MockLink {
  constructor() {
    super([]);
  }

  request() {
    return null;
  }
}

export const fakeContextProgram = {
  selectedProgram: {
    id: '1',
    name: 'someName',
    status: ProgramStatus.Active,
    pduFields: null,
    dataCollectingType: {
      id: '1',
      householdFiltersAvailable: true,
      individualFiltersAvailable: true,
      label: 'data collecting type',
      code: '123',
      type: 'full',
      children: null,
    },
  },
  setSelectedProgram: () => {},
  isActiveProgram: true,
  isSocialDctType: false,
  isStandardDctType: true,
  programHasPdu: false,
};
