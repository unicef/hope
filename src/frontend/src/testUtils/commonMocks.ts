import { vi } from 'vitest';
import React from 'react';

/**
 * Common mock for @hooks/useBaseUrl hook
 */
export const mockUseBaseUrl = (
  businessArea = 'afghanistan',
  programId = 'test-program',
  baseUrl = 'afghanistan/test-program',
) => {
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({ businessArea, programId, baseUrl }),
  }));
};

/**
 * Common mock for useProgramContext hook
 */
export const mockUseProgramContext = (
  groupLabel = 'Household',
  groupLabelPlural = 'Households',
) => {
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel,
          groupLabelPlural,
        },
      },
    }),
  }));
};

/**
 * Common mock for react-router-dom
 */
export const mockReactRouterDom = () => {
  vi.mock('react-router-dom', () => ({
    useNavigate: () => vi.fn(),
    useLocation: () => ({
      pathname: '/test-path',
      search: '',
      hash: '',
      state: null,
      key: 'test-key',
    }),
    Link: ({ children, to, ...props }: any) =>
      React.createElement('a', { href: to, ...props }, children),
  }));
};
/**
 * Common mock for @utils/utils module
 */
export const utilsMock = {
  createApiParams: vi.fn((baseParams, queryParams, options) => ({
    ...baseParams,
    ...queryParams,
    ...options,
  })),
  adjustHeadCells: vi.fn((headCells) => headCells),
  formatCurrencyWithSymbol: vi.fn(
    (amount, currency) => `${amount} ${currency}`,
  ),
  householdStatusToColor: vi.fn(() => '#000000'),
  filterEmptyParams: vi.fn((params) => {
    if (!params) return {};
    return Object.fromEntries(
      Object.entries(params).filter(([, value]) => {
        return value !== undefined && value !== null && value !== '';
      }),
    );
  }),
  opacityToHex: vi.fn((opacity) =>
    Math.round(opacity * 255)
      .toString(16)
      .padStart(2, '0'),
  ),
  isPermissionDeniedError: vi.fn((error) => error?.response?.status === 403),
  dateToIsoString: vi.fn((date) => date?.toISOString?.() || date),
  choicesToDict: vi.fn((choices) => {
    if (!choices) return {};
    return choices.reduce((acc, choice) => {
      acc[choice.value] = choice.name;
      return acc;
    }, {});
  }),
  populationStatusToColor: vi.fn(() => 'primary'),
  sexToCapitalize: vi.fn(
    (sex) => sex?.charAt(0).toUpperCase() + sex?.slice(1).toLowerCase(),
  ),
  programStatusToColor: vi.fn(() => 'primary'),
  formatCurrency: vi.fn((amount) => `$${amount?.toLocaleString() || '0'}`),
  columnToOrderBy: vi.fn((column) => column || 'name'),
};

/**
 * Helper function to mock the @utils/utils module
 */
export const mockUtils = () => {
  vi.mock('@utils/utils', () => utilsMock);
};

/**
 * Common mock for RestService
 */
export const mockRestService = () => {
  vi.mock('@restgenerated/services/RestService', () => ({
    RestService: {
      restBusinessAreasProgramsHouseholdsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsCountRetrieve: vi.fn(),
      restBusinessAreasProgramsIndividualsList: vi.fn(),
      restBusinessAreasProgramsHouseholdsMembersList: vi.fn(),
      restBusinessAreasProgramsList: vi.fn(),
      restBusinessAreasProgramsCountRetrieve: vi.fn(),
      restBusinessAreasUsersProfileRetrieve: vi.fn(),
      // Add other commonly used RestService methods as needed
    },
  }));
};

/**
 * Helper function to set up all common mocks at once
 * Usage in test files:
 *
 * import { setupCommonMocks } from 'src/testUtils/commonMocks';
 *
 * // At the top level of your test file
 * setupCommonMocks();
 */
export const setupCommonMocks = () => {
  // Mock useBaseUrl
  vi.mock('@hooks/useBaseUrl', () => ({
    useBaseUrl: () => ({
      businessArea: 'afghanistan',
      programId: 'test-program',
      baseUrl: 'afghanistan/test-program',
    }),
  }));

  // Mock useProgramContext
  vi.mock('src/programContext', () => ({
    useProgramContext: () => ({
      selectedProgram: {
        beneficiaryGroup: {
          groupLabel: 'Household',
          groupLabelPlural: 'Households',
        },
      },
    }),
  }));

  mockReactRouterDom();
  mockUtils();
  mockRestService();
};
