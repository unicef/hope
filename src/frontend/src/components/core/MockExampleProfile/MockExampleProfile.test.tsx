import { expect, it } from 'vitest';
import MockExampleProfile from './MockExampleProfile'; // Adjust the import path as needed
import { renderWithProviders } from '../../../testUtils/testUtils';
import { waitFor, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

it('MockExampleProfile renders correctly after data is fetched', async () => {
  const { asFragment } = renderWithProviders(<MockExampleProfile />);

  // Wait for an element that indicates the data has been loaded
  await waitFor(() => {
    expect(screen.getByText(/Profile Details/i)).toBeInTheDocument();
  });

  // Take the snapshot after ensuring the data is loaded
  expect(asFragment()).toMatchSnapshot();
});
