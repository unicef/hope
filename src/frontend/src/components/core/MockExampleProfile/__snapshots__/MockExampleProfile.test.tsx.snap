// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MockExampleProfile renders correctly after data is fetched 1`] = `
<DocumentFragment>
  <div
    style="border: 3px solid black; padding: 10px; width: 200px;"
  >
    <div>
      <h4>
        Profile Details
      </h4>
      Root Rootkowski
    </div>
  </div>
</DocumentFragment>
`;
